pytket/extensions/cutensornet/mps/mps.py

```diff
@@ -721,23 +737,37 @@ def get_physical_dimension(self, position: int) -> int:
             self.get_physical_bond(position)
         )
 
-    def copy(self, device_id: Optional[int] = None) -> MPS:
-        """Returns a deep copy of self.
+    def get_device_id(self) -> int:
+        """
+        Returns:
+            The identifier of the device (GPU) where the tensors are stored.
+        """
+        return int(self.tensors[0].data.device)
 
-        Args:
-            device_id: The identifier of the GPU where the copy is meant to be kept.
-                If not provided, the same device as ``self`` will be used.
+    def set_libhandle(self, libhandle: CuTensorNetHandle) -> None:
+        """Set the library handle used by this ``MPS`` object. Multiple objects
+        may use the same library handle.
+
+        Raises:
+            RuntimeError: If the device (GPU) where ``libhandle`` was initialised
+                does not match the one where the tensors of the MPS are stored.
+        """
+        if libhandle.device_id != self.get_device_id():
+            raise RuntimeError(
+                "Device of libhandle is not the one where the MPS is stored.",
+                f"{libhandle.device_id} != {self.get_device_id()}",
+            )
+        self._lib = libhandle
 
+    def copy(self) -> MPS:
+        """
         Returns:
-            A deep copy of the MPS.
+            A deep copy of the MPS on the same device.
         """
         self._flush()
 
-        if device_id is None:
-            device_id = self._device_id
-
         # Create a dummy object
-        new_mps = MPS(qubits=[], chi=self.chi, device_id=device_id)
+        new_mps = MPS(qubits=[], chi=self.chi, device_id=self.get_device_id())

```

Comment:
The reason for this change was that I am not confident that I can safely copy from one device to another just using CuPy. For now, I am disallowing it, and we can come back to this later on if necessary.

In any case, I have checked that `bcast` from `mpi4py` does copy data between devices (at least in Perlmutter with CUDA-aware MPI is set up properly), so that's probably the way you'd want to do it anyway.